File: app/src/main/java/com/amazon/tv/leanbacklauncher/HomeScrollManager.java
package com.amazon.tv.leanbacklauncher;

import android.content.Context;
import android.content.res.Resources;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.amazon.tv.leanbacklauncher.animation.ParticipatesInScrollAnimation;

import java.util.ArrayList;

public class HomeScrollManager {
    private boolean mAnimationsEnabled = true;
    private float mFractionFromTop;
    private final RecyclerView mRecyclerView;
    private int mScrollAnimationThreshold;
    private ArrayList<HomeScrollFractionListener> mScrollListeners = new ArrayList();
    private int mScrollPosition;
    private int mScrollThreshold;

    public interface HomeScrollFractionListener {
        void onScrollPositionChanged(int i, float f);
    }

    public HomeScrollManager(Context context, RecyclerView recyclerView) {
        this.mRecyclerView = recyclerView;
        Resources resources = context.getResources();
        this.mScrollThreshold = resources.getDimensionPixelOffset(R.dimen.home_scroll_size_search);
        this.mScrollAnimationThreshold = resources.getDimensionPixelOffset(R.dimen.home_scroll_animation_threshold);
    }

    public void addHomeScrollListener(HomeScrollFractionListener listener) {
        if (!this.mScrollListeners.contains(listener)) {
            this.mScrollListeners.add(listener);
            listener.onScrollPositionChanged(this.mScrollPosition, this.mFractionFromTop);
        }
    }

    public void removeHomeScrollListener(HomeScrollFractionListener listener) {
        for (int i = 0; i < this.mScrollListeners.size(); i++) {
            if (this.mScrollListeners.get(i) == listener) {
                this.mScrollListeners.remove(i);
                return;
            }
        }
    }

    public void onScrolled(int dy, int scrollPosition) {
        if (this.mScrollPosition != scrollPosition) {
            this.mScrollPosition = scrollPosition;
            if (this.mScrollThreshold > 0) {
                this.mFractionFromTop = Math.max(0.0f, Math.min(1.0f, Math.abs(((float) this.mScrollPosition) / ((float) this.mScrollThreshold))));
            } else {
                this.mFractionFromTop = 0.0f;
            }
            boolean shouldAnimate = Math.abs(dy) <= this.mScrollAnimationThreshold;
            if (shouldAnimate != this.mAnimationsEnabled) {
                this.mAnimationsEnabled = shouldAnimate;
                adjustAnimationsEnabledState(this.mRecyclerView);
            }
            updateListeners();
        }
    }

    private void updateListeners() {
        for (int i = 0; i < this.mScrollListeners.size(); i++) {
            this.mScrollListeners.get(i).onScrollPositionChanged(this.mScrollPosition, this.mFractionFromTop);
        }
    }

    public void onScrollStateChanged(int newState) {
        this.mAnimationsEnabled = newState == 0;
        adjustAnimationsEnabledState(this.mRecyclerView);
    }

    private void adjustAnimationsEnabledState(View view) {
        if (view instanceof ParticipatesInScrollAnimation) {
            ((ParticipatesInScrollAnimation) view).setAnimationsEnabled(this.mAnimationsEnabled);
        }
        if (view instanceof ViewGroup) {
            int count = ((ViewGroup) view).getChildCount();
            for (int i = 0; i < count; i++) {
                adjustAnimationsEnabledState(((ViewGroup) view).getChildAt(i));
            }
        }
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/HomeScrollManager.kt
package com.amazon.tv.leanbacklauncher

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.amazon.tv.leanbacklauncher.animation.ParticipatesInScrollAnimation
import java.util.*

class HomeScrollManager(context: Context, private val mRecyclerView: RecyclerView) {
    private var mAnimationsEnabled = true
    private var mFractionFromTop = 0f
    private val mScrollAnimationThreshold: Int
    private val mScrollListeners: ArrayList<HomeScrollFractionListener?> = arrayListOf()
    private var mScrollPosition = 0
    private val mScrollThreshold: Int

    init {
        val resources = context.resources
        mScrollThreshold = resources.getDimensionPixelOffset(R.dimen.home_scroll_size_search)
        mScrollAnimationThreshold = resources.getDimensionPixelOffset(R.dimen.home_scroll_animation_threshold)
    }

    interface HomeScrollFractionListener {
        fun onScrollPositionChanged(i: Int, f: Float)
    }

    fun addHomeScrollListener(listener: HomeScrollFractionListener) {
        if (!mScrollListeners.contains(listener)) {
            mScrollListeners.add(listener)
            listener.onScrollPositionChanged(mScrollPosition, mFractionFromTop)
        }
    }

    fun removeHomeScrollListener(listener: HomeScrollFractionListener) {
        for (i in mScrollListeners.indices) {
            if (mScrollListeners[i] === listener) {
                mScrollListeners.removeAt(i)
                return
            }
        }
    }

    fun onScrolled(dy: Int, scrollPosition: Int) {
        if (mScrollPosition != scrollPosition) {
            mScrollPosition = scrollPosition
            if (mScrollThreshold > 0) {
                mFractionFromTop = Math.max(0.0f, Math.min(1.0f, Math.abs(mScrollPosition.toFloat() / mScrollThreshold.toFloat())))
            } else {
                mFractionFromTop = 0.0f
            }
            val shouldAnimate = Math.abs(dy) <= mScrollAnimationThreshold
            if (shouldAnimate != mAnimationsEnabled) {
                mAnimationsEnabled = shouldAnimate
                adjustAnimationsEnabledState(mRecyclerView)
            }
            updateListeners()
        }
    }

    private fun updateListeners() {
        for (i in mScrollListeners.indices) {
            mScrollListeners[i]!!.onScrollPositionChanged(mScrollPosition, mFractionFromTop)
        }
    }

    fun onScrollStateChanged(newState: Int) {
        mAnimationsEnabled = newState == 0
        adjustAnimationsEnabledState(mRecyclerView)
    }

    private fun adjustAnimationsEnabledState(view: View) {
        if (view is ParticipatesInScrollAnimation) {
            (view as ParticipatesInScrollAnimation).setAnimationsEnabled(mAnimationsEnabled)
        }
        if (view is ViewGroup) {
            val count = view.childCount
            for (i in 0 until count) {
                adjustAnimationsEnabledState(view.getChildAt(i))
            }
        }
    }
}